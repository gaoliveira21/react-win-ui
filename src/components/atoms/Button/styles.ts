import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from '.'

const modifiers = {
  primary: (theme: DefaultTheme) => css`
    color: ${theme.colors.text.ON_ACCENT};
    background-color: ${theme.colors.fill.accent.ACCENT_PRIMARY};

    &:hover {
      background-color: ${theme.colors.fill.accent.ACCENT_SECONDARY};
    }

    &:active {
      background-color: ${theme.colors.fill.accent.ACCENT_TERTIARY};
    }

    &:disabled {
      color: ${theme.colors.text.ON_ACCENT_DISABLED};
      background-color: ${theme.colors.fill.accent.DISABLED};
    }
  `,

  secondary: (theme: DefaultTheme) => css`
    color: ${theme.colors.text.PRIMARY};
    background-color: ${theme.colors.fill.control.CONTROL_PRIMARY};

    &:hover {
      background-color: ${theme.colors.fill.control.CONTROL_SECONDARY};
    }

    &:active {
      background-color: ${theme.colors.fill.control.CONTROL_TERTIARY};
    }

    &:disabled {
      color: ${theme.colors.text.DISABLED};
      background-color: ${theme.colors.fill.control.CONTROL_PRIMARY};
    }
  `
}

export const Button = styled.button<Pick<ButtonProps, 'variant'>>`
  ${({ theme, variant }) => css`
    font-size: ${theme.fonts.text.BODY.size};
    font-weight: ${theme.fonts.text.BODY.weight};
    line-height: ${theme.fonts.text.BODY.lineHeight};

    box-shadow: 0 0 2px 2px ${theme.colors.elevation.control.BORDER};

    padding: 5px 47px;
    border-radius: 4px;

    transition: background-color 0.3s;

    &:disabled {
      cursor: not-allowed;
    }

    &:focus {
      outline: 3px solid ${theme.colors.stroke.focus.OUTER};
      border: 1px solid ${theme.colors.stroke.focus.INNER};
    }

    ${!!variant && modifiers[variant](theme)};
  `}
`
