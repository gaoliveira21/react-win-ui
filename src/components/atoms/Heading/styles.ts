import styled, { css } from 'styled-components'

import { HeadingProps } from '.'

export const Heading = styled.div<Pick<HeadingProps, 'variant' | 'color'>>`
  ${({ theme, variant, color }) => css`
  font-family: ${theme.fonts.heading[variant.toUpperCase()].fontFamily};}
    font-size: ${theme.fonts.heading[variant.toUpperCase()].size};}
    font-weight: ${theme.fonts.heading[variant.toUpperCase()].weight};}
    line-height: ${theme.fonts.heading[variant.toUpperCase()].lineHeight};}
    color: ${theme.colors.text[color.toUpperCase()]};
  `}
`
