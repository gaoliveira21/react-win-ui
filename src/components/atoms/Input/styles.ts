import styled, { css } from 'styled-components'

export const Input = styled.input`
  ${({ theme }) => css`
    background-color: ${theme.colors.fill.control.CONTROL_PRIMARY};
    border: 1px solid ${theme.colors.elevation.textControl.BORDER};
    border-bottom: 2px solid ${theme.colors.stroke.control.SECONDARY};

    color: ${theme.colors.text.PRIMARY};
    font-size: ${theme.fonts.text.BODY.size};
    font-weight: ${theme.fonts.text.BODY.weight};
    line-height: ${theme.fonts.text.BODY.lineHeight};

    padding: 4px 35px 6px 11px;
    border-radius: 3px;

    &::placeholder {
      font-size: ${theme.fonts.text.BODY.size};
      font-weight: ${theme.fonts.text.BODY.weight};
      line-height: ${theme.fonts.text.BODY.lineHeight};
      color: ${theme.colors.text.SECONDARY};
    }

    &:focus {
      outline: none;
      border-bottom: 2px solid ${theme.colors.fill.accent.ACCENT_PRIMARY};
      background-color: ${theme.colors.fill.control.ACTIVE};
    }

    &:hover {
      &:not(:focus, :disabled) {
        background-color: ${theme.colors.fill.control.CONTROL_SECONDARY};
      }
    }

    &:disabled {
      background-color: ${theme.colors.fill.control.DISABLED};
      color: ${theme.colors.text.DISABLED};
      cursor: not-allowed;
    }
  `}
`
