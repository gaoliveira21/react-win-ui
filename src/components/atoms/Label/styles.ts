import styled, { css } from 'styled-components'

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.fonts.text.BODY.size};
    font-weight: ${theme.fonts.text.BODY.weight};
    line-height: ${theme.fonts.text.BODY.lineHeight};
    color: ${theme.colors.text.PRIMARY};
  `}
`
