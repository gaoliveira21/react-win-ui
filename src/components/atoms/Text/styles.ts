import styled from 'styled-components'

import { TextProps } from '.'

export const Text = styled.p<Pick<TextProps, 'variant'>>`
  font-family: ${({ theme, variant }) =>
    theme.fonts.text[variant.toUpperCase()].fontFamily};
  font-size: ${({ theme, variant }) =>
    theme.fonts.text[variant.toUpperCase()].size};
  font-weight: ${({ theme, variant }) =>
    theme.fonts.text[variant.toUpperCase()].weight};
  line-height: ${({ theme, variant }) =>
    theme.fonts.text[variant.toUpperCase()].lineHeight};
`
