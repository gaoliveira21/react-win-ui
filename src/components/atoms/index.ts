export * from './Text'
export * from './Heading'
export * from './Button'
export * from './Input'
export * from './Label'
