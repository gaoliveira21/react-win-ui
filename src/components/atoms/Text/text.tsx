import { TextTypes } from '@/styles/colors'
import { BodyFonts } from '@/styles/typography'

import * as S from './styles'

export type TextProps = {
  children: React.ReactNode
  as?: 'p' | 'strong' | 'span' | 'output' | 'i' | 'small'
  variant?: Lowercase<keyof BodyFonts>
  color?: Lowercase<TextTypes>
}

export function Text({
  children,
  as = 'p',
  variant = 'body',
  color = 'primary'
}: TextProps): React.ReactElement {
  return (
    <S.Text as={as} variant={variant} color={color}>
      {children}
    </S.Text>
  )
}
