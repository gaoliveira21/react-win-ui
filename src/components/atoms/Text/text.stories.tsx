import { Meta, Story } from '@storybook/react/types-6-0'

import { Text } from './text'

export default {
  title: 'Text',
  component: Text
} as Meta

export const Basic: Story = (args) => (
  <Text {...args}>Basic Text Component</Text>
)
