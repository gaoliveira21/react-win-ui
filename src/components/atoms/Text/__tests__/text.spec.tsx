import { screen } from '@testing-library/react'

import { bodyFonts } from '@/styles/typography'
import { ligthTextColors, darkTextColors } from '@/styles/colors'

import { renderWithTheme } from '@/tests/render-with-theme'

import { Text, TextProps } from '@/components/atoms/Text'

const assertFontStyle = (sut: Element, font: TextProps['variant'] = 'body') => {
  expect(sut).toHaveStyle({
    'font-family': bodyFonts[font.toUpperCase()].fontFamily,
    'font-size': bodyFonts[font.toUpperCase()].size,
    'line-height': bodyFonts[font.toUpperCase()].lineHeight,
    'font-weight': bodyFonts[font.toUpperCase()].weight,
    color: ligthTextColors.PRIMARY
  })
}

describe('<Text />', () => {
  it('Should render Text with correct initial styles', () => {
    renderWithTheme(<Text>text_content</Text>)

    assertFontStyle(screen.getByText(/text_content/i))
  })

  it('Should render Text with correct children prop', () => {
    const textContent = 'text_content'
    renderWithTheme(<Text>{textContent}</Text>)

    expect(screen.getByText(textContent)).toBeInTheDocument()
  })

  it('Should render Text as paragraph by default', () => {
    renderWithTheme(<Text>text_content</Text>)

    expect(screen.getByText(/text_content/i).tagName.toLowerCase()).toBe('p')
  })

  it('Should render Text with different html tag if "as" prop has been passed', () => {
    const tagName = 'strong'
    renderWithTheme(<Text as={tagName}>text_content</Text>)

    expect(screen.getByText(/text_content/i).tagName.toLowerCase()).toBe(
      tagName
    )
  })

  it('Should render with "BODY_LARGE" styles if "variants" prop is equal to "body_large"', () => {
    renderWithTheme(<Text variant="body_large">text_content</Text>)

    assertFontStyle(screen.getByText(/text_content/i), 'body_large')
  })

  it('Should render with "CAPTION" styles if "variants" prop is equal to "caption"', () => {
    renderWithTheme(<Text variant="caption">text_content</Text>)

    assertFontStyle(screen.getByText(/text_content/i), 'caption')
  })

  it('Should render with "BODY_STRONG" styles if "variants" prop is equal to "body_strong"', () => {
    renderWithTheme(<Text variant="body_strong">text_content</Text>)

    assertFontStyle(screen.getByText(/text_content/i), 'body_strong')
  })

  it('Should render with correct color', () => {
    renderWithTheme(<Text color="accent_primary">text_content</Text>)

    expect(screen.getByText(/text_content/i)).toHaveStyle({
      color: ligthTextColors.ACCENT_PRIMARY
    })
  })

  it('Should render with correct color if dark mode is enable', () => {
    renderWithTheme(<Text>text_content</Text>, { dark: true })

    expect(screen.getByText(/text_content/i)).toHaveStyleRule(
      'color',
      darkTextColors.PRIMARY
    )
  })
})
