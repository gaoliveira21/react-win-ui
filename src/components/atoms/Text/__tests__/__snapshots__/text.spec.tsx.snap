// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> Should render Text with correct initial styles 1`] = `
.c0 {
  font-family: SegoeUI;
  font-size: 14px;
  font-weight: 400;
  line-height: 20px;
  color: rgb(0,0,0);
}

<p
  class="c0"
  color="primary"
>
  text_content
</p>
`;
